app/src/androidTest/java/com/github/sdpteam15/polyevents/view/activity/EventActivityTest.kt

```diff
@@ -246,18 +259,233 @@ class EventActivityTest {
+
+        // Click review event
+        clickOn(R.id.event_leave_review_button)
+        assertDisplayed(R.id.leave_review_dialog_fragment)
+
+        onView(withId(R.id.leave_review_fragment_rating)).perform(SetRating(4.0f))
+        clickOn(R.id.leave_review_fragment_save_button)
+
+        //assertNull(createdRating?.feedback)
+        assertEquals(createdRating?.rate, 4.0f)
+        assertEquals(createdRating?.userId, testUser.uid)
+        assertEquals(createdRating?.eventId, limitedEventId)
+    }
+
+    @Test
+    fun testUserCanCancelLeavingRating() {
+        var createdRating: Rating? = null
+        When(mockedEventDatabase.addRatingToEvent(
+            anyOrNull(), anyOrNull()
+        )).thenAnswer {
+            createdRating = (it.arguments[0] as Rating?)
+            Observable(true)
+        }
+
+        goToEventActivityWithIntent()
+
+        // Click review event
+        clickOn(R.id.event_leave_review_button)
+        assertDisplayed(R.id.leave_review_dialog_fragment)
+
+        onView(withId(R.id.leave_review_fragment_rating)).perform(SetRating(4.0f))
+        clickOn(R.id.leave_review_fragment_cancel_button)
+
+        assertNull(createdRating)
+        assertNotExist(R.id.leave_review_dialog_fragment)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    @Test
+    fun testShouldUpdateRatingIfUserRatedAlready() {
+        var existingRating = Rating(
+            ratingId = "1",
+            userId = testUser.uid,
+            eventId = limitedEventId,
+            rate = 4.0f,
+            feedback = "Noice"
+        )
+        When(mockedEventDatabase.getUserRatingFromEvent(
+            userId = anyOrNull(),
+            eventId = anyOrNull(),
+            returnedRating = anyOrNull(),
+            userAccess = anyOrNull()
+        )).thenAnswer {
+            // Not very robust, need to change if changed method signature
+            (it.arguments[2] as Observable<Rating>).postValue(
+                existingRating
+            )
+            Observable(true)
+        }
+
+        When(mockedEventDatabase.updateRating(
+            rating = anyOrNull(),
+            userAccess = anyOrNull()
+        )).then {
+            existingRating = it.arguments[0] as Rating
+            Observable(true)
+        }
+
+        goToEventActivityWithIntent()
+
+        // Click review event
+        clickOn(R.id.event_leave_review_button)
+        assertDisplayed(R.id.leave_review_dialog_fragment)
+
+        // Check fetched rating is displayed
+        assertDisplayed(R.id.leave_review_fragment_feedback_text, existingRating.feedback!!)
+        assertProgress(R.id.leave_review_fragment_rating, existingRating.rate!!.toInt())
+
+        onView(withId(R.id.leave_review_fragment_rating)).perform(SetRating(3.0f))
+        clickOn(R.id.leave_review_fragment_save_button)
+
+        assertNotExist(R.id.leave_review_dialog_fragment)
+        assertEquals(existingRating.rate, 3.0f)
+    }
+
+    @Suppress("UNCHECKED_CAST")
```

Comment:
You can put the @Suppress("UNCHECKED_CAST") on the class so that you don't have to put it on each method

Reply:
Yes but as I said in your PR, I'm keeping them local to concerned function methods, in case I have any actual bad casts elsewhere in the class